File: packages/omnia/src/transition.ts
import type { ProgramTransition } from './types.js';

type TransitionScalar = string | bigint | boolean;

function canonicalScalar(value: unknown, path: string): TransitionScalar {
  if (typeof value === 'string' || typeof value === 'bigint' || typeof value === 'boolean') return value;
  throw new Error(`Invalid ProgramTransition ${path}: expected string, bigint, or boolean`);
}

function canonicalRecord<T extends TransitionScalar>(
  record: Record<string, unknown> | undefined,
  path: string,
): Record<string, T> | undefined {
  if (record === undefined) return undefined;
  if (record === null || typeof record !== 'object' || Array.isArray(record)) {
    throw new Error(`Invalid ProgramTransition ${path}: expected object`);
  }
  const out: Record<string, T> = {};
  for (const key of Object.keys(record).sort()) {
    out[key] = canonicalScalar(record[key], `${path}.${key}`) as T;
  }
  return out;
}

function canonicalStringRecord(
  record: Record<string, unknown> | undefined,
  path: string,
): Record<string, string> | undefined {
  if (record === undefined) return undefined;
  if (record === null || typeof record !== 'object' || Array.isArray(record)) {
    throw new Error(`Invalid ProgramTransition ${path}: expected object`);
  }
  const out: Record<string, string> = {};
  for (const key of Object.keys(record).sort()) {
    const value = record[key];
    if (typeof value !== 'string') {
      throw new Error(`Invalid ProgramTransition ${path}.${key}: expected string`);
    }
    out[key] = value;
  }
  return out;
}

export function canonicalizeProgramTransition(
  transition?: ProgramTransition,
): ProgramTransition | undefined {
  if (transition === undefined) return undefined;
  if (!transition || typeof transition !== 'object' || Array.isArray(transition)) {
    throw new Error('Invalid ProgramTransition: expected object');
  }
  if (typeof transition.action !== 'string' || transition.action.length === 0) {
    throw new Error('Invalid ProgramTransition action: expected non-empty string');
  }

  const canonical: ProgramTransition = { action: transition.action };
  const inputs = canonicalRecord(transition.inputs, 'inputs');
  if (inputs && Object.keys(inputs).length > 0) canonical.inputs = inputs;
  const witness = canonicalStringRecord(transition.witness, 'witness');
  if (witness && Object.keys(witness).length > 0) canonical.witness = witness;
  const metadata = canonicalStringRecord(transition.metadata, 'metadata');
  if (metadata && Object.keys(metadata).length > 0) canonical.metadata = metadata;
  return canonical;
}

function canonicalJsonValue(value: unknown): string {
  if (typeof value === 'bigint') return JSON.stringify({ __bigint: value.toString() });
  if (value === null || typeof value !== 'object') return JSON.stringify(value);
  if (Array.isArray(value)) return `[${value.map(canonicalJsonValue).join(',')}]`;
  const obj = value as Record<string, unknown>;
  return `{${Object.keys(obj).sort().map(k => `${JSON.stringify(k)}:${canonicalJsonValue(obj[k])}`).join(',')}}`;
}

export function serializeProgramTransition(transition: ProgramTransition): string {
  return canonicalJsonValue(canonicalizeProgramTransition(transition));
}

File: packages/omnia/rust/src/transition.rs
use serde_json::{Map, Value};

use crate::types::ProgramTransition;

fn canonical_scalar(value: &Value, path: &str) -> Result<Value, String> {
    match value {
        Value::String(_) | Value::Bool(_) => Ok(value.clone()),
        _ => Err(format!(
            "Invalid ProgramTransition {}: expected string, bigint, or boolean",
            path
        )),
    }
}

fn canonical_record(record: Option<&Value>, path: &str) -> Result<Option<Value>, String> {
    let Some(record) = record else {
        return Ok(None);
    };
    let Value::Object(obj) = record else {
        return Err(format!(
            "Invalid ProgramTransition {}: expected object",
            path
        ));
    };
    if obj.is_empty() {
        return Ok(None);
    }

    let mut out = Map::new();
    let mut keys: Vec<&String> = obj.keys().collect();
    keys.sort();
    for key in keys {
        out.insert(
            key.clone(),
            canonical_scalar(&obj[key], &format!("{}.{}", path, key))?,
        );
    }
    Ok(Some(Value::Object(out)))
}

fn canonical_string_record(record: Option<&Value>, path: &str) -> Result<Option<Value>, String> {
    let Some(record) = record else {
        return Ok(None);
    };
    let Value::Object(obj) = record else {
        return Err(format!(
            "Invalid ProgramTransition {}: expected object",
            path
        ));
    };
    if obj.is_empty() {
        return Ok(None);
    }

    let mut out = Map::new();
    let mut keys: Vec<&String> = obj.keys().collect();
    keys.sort();
    for key in keys {
        match &obj[key] {
            Value::String(value) => {
                out.insert(key.clone(), Value::String(value.clone()));
            }
            _ => {
                return Err(format!(
                    "Invalid ProgramTransition {}.{}: expected string",
                    path, key
                ));
            }
        }
    }
    Ok(Some(Value::Object(out)))
}

pub fn canonicalize_program_transition(
    transition: Option<&ProgramTransition>,
) -> Result<Option<ProgramTransition>, String> {
    let Some(transition) = transition else {
        return Ok(None);
    };
    if transition.action.is_empty() {
        return Err("Invalid ProgramTransition action: expected non-empty string".to_string());
    }

    Ok(Some(ProgramTransition {
        action: transition.action.clone(),
        inputs: canonical_record(transition.inputs.as_ref(), "inputs")?,
        witness: canonical_string_record(transition.witness.as_ref(), "witness")?,
        metadata: canonical_string_record(transition.metadata.as_ref(), "metadata")?,
    }))
}

pub fn validate_program_transition_shape(transition: &ProgramTransition) -> Result<(), String> {
    canonicalize_program_transition(Some(transition)).map(|_| ())
}

fn canonical_json_value(value: &Value) -> String {
    match value {
        Value::Array(values) => format!(
            "[{}]",
            values
                .iter()
                .map(canonical_json_value)
                .collect::<Vec<_>>()
                .join(",")
        ),
        Value::Object(obj) => {
            let mut keys: Vec<&String> = obj.keys().collect();
            keys.sort();
            let body = keys
                .iter()
                .map(|key| {
                    format!(
                        "{}:{}",
                        serde_json::to_string(key).unwrap_or_default(),
                        canonical_json_value(&obj[*key])
                    )
                })
                .collect::<Vec<_>>()
                .join(",");
            format!("{{{}}}", body)
        }
        _ => serde_json::to_string(value).unwrap_or_default(),
    }
}

pub fn serialize_program_transition(transition: &ProgramTransition) -> Result<String, String> {
    let canonical = canonicalize_program_transition(Some(transition))?
        .ok_or_else(|| "Invalid ProgramTransition: expected object".to_string())?;
    let value = serde_json::to_value(canonical).map_err(|e| e.to_string())?;
    Ok(canonical_json_value(&value))
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::json;

    #[test]
    fn serializes_transition_with_sorted_fields_and_records() {
        let transition = ProgramTransition {
            action: "set".to_string(),
            inputs: Some(json!({ "value": "42", "enabled": true })),
            witness: Some(json!({ "source": "sensor", "proof": "0xabc" })),
            metadata: Some(json!({ "requestId": "req-1" })),
        };

        assert_eq!(
            serialize_program_transition(&transition).unwrap(),
            "{\"action\":\"set\",\"inputs\":{\"enabled\":true,\"value\":\"42\"},\"metadata\":{\"requestId\":\"req-1\"},\"witness\":{\"proof\":\"0xabc\",\"source\":\"sensor\"}}"
        );
    }

    #[test]
    fn omits_empty_optional_records() {
        let transition = ProgramTransition {
            action: "increment".to_string(),
            inputs: Some(json!({})),
            witness: None,
            metadata: Some(json!({})),
        };

        assert_eq!(
            serialize_program_transition(&transition).unwrap(),
            "{\"action\":\"increment\"}"
        );
    }

    #[test]
    fn rejects_invalid_transition_shapes() {
        let empty_action = ProgramTransition {
            action: "".to_string(),
            inputs: None,
            witness: None,
            metadata: None,
        };
        assert_eq!(
            serialize_program_transition(&empty_action).unwrap_err(),
            "Invalid ProgramTransition action: expected non-empty string"
        );

        let numeric_input = ProgramTransition {
            action: "set".to_string(),
            inputs: Some(json!({ "value": 42 })),
            witness: None,
            metadata: None,
        };
        assert_eq!(
            serialize_program_transition(&numeric_input).unwrap_err(),
            "Invalid ProgramTransition inputs.value: expected string, bigint, or boolean"
        );

        let non_string_witness = ProgramTransition {
            action: "set".to_string(),
            inputs: None,
            witness: Some(json!({ "proof": true })),
            metadata: None,
        };
        assert_eq!(
            serialize_program_transition(&non_string_witness).unwrap_err(),
            "Invalid ProgramTransition witness.proof: expected string"
        );
    }
}
